wled00/util.cpp

```diff
@@ -0,0 +1,33 @@
+#include "wled.h"
+#include "fcn_declare.h"
+#include "const.h"
+
+
+bool requestJSONBufferLock(uint8_t module)
+{
+  unsigned long now = millis();
+
+  while (jsonBufferLock && millis()-now < 1000) delay(1); // wait for a second for buffer lock
```

Comment:
I thought to myself that this can't possibly work, because `delay()` in ESP8266 network callbacks will crash and burn.

Turns out that this _does_ in fact work, because ESP8266 and 32 differ in behavior:
ESP8266 would be stuck for a second and not reply to the request in the best case and crash in the worst case if it encountered a locked buffer in a network callback. Luckily this _should_ never happen as the ESP8266 can only handle network callbacks wherever `yield()` is placed. As such, as long as we make sure we never call `yield()` while the buffer is locked, we are 100% fine. In fact, buffer lock logic is _not required at all_ on ESP8266 since there is simply no way for two interactions that both need the buffer to intersect.

On ESP32, the buffer lock _is_ needed and works as intended! Due to being dual core, network callbacks are run on core 1 while the main loop is run on core 0. If core 1 handles a network request, it will wait out the lock, as (unlike with 8266) the loop task will continue to run, thus only blocking for a few milliseconds at most.

I'm interested in how ESP32-S2 would behave here, given that it is single core. I would assume same as ESP8266, but it might also be running two tasks on the same core (and rapidly switching between them).
In both cases, it _should_ work fine.

I think this is ready for merge and public testing as part of 0.13.0-b6 :)

Reply:
Never thought about implications of net callbacks since I was unaware of the limitations.
Still, this code is now running bug free for a month on a ESP01, D1 mini and D1 min32.
I went so far as to install it on all of my production units. Not a single one of them has crashed since!